### ISSUE TITLE
ESLint gives false error in for .. of loop with Maps

### ISSUE THREAD

**Issue by brizental** [eslint/eslint#9732]
* **ESLint Version:**
4.12.1
* **Node Version:**
9.2.0
* **npm Version:**
5.6.0

**What parser (default, Babel-ESLint, etc.) are you using?**
default

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
module.exports = {
    "env": {
        "browser": true,
        "commonjs": true,
        "es6": true
    },
    "extends": "eslint:recommended",
    "parserOptions": {
        "sourceType": "module"
    },
    "plugins": [
        "prettier",
        "json"
    ],
    "rules": {
        "prettier/prettier": [
            "error",
            {
                "tabWidth": 2,
                "singleQuote": false
            }
        ],
        "indent": [
            "error",
            2
        ],
        "linebreak-style": [
            "error",
            "unix"
        ],
        "quotes": [
            "error",
            "double"
        ],
        "semi": [
            "error",
            "always"
        ]
    }
};
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
   for ([index, point] of points.entries()) {
      if (point.x === this.point.x - this.offsetX &&
          point.y === this.point.y - this.offsetY &&
          this.node.id === "") {
        this.node.id = "direction-" + index;
      }
    }
```

```bash
./node_modules/.bin/eslint ./lib/js/*.js
```

**What did you expect to happen?**
I expected it to understand the syntax and pass.

**What actually happened? Please include the actual, raw output from ESLint.**
```bash
  29:11  error  'index' is not defined  no-undef
  29:18  error  'point' is not defined  no-undef
  31:9   error  'point' is not defined  no-undef
  32:9   error  'point' is not defined  no-undef
  35:39  error  'index' is not defined  no-undef

✖ 5 problems (5 errors, 0 warnings)
```

**Comment by gyandeeps** [comment]
I think the rule working as expected. In your code you should use `const` to declare the destructed variables.

```js
    for (const [index, point] of points.entries()) {
      if (point.x === this.point.x - this.offsetX &&
          point.y === this.point.y - this.offsetY &&
          this.node.id === "") {
        this.node.id = "direction-" + index;
      }
    }
```

**Comment by brizental** [comment]
Yeah! Tested and indeed that was the problem. Thanks!
